### PR TITLE
feat: Flat fare priced by distance Netex

### PR DESCRIPTION
**Pull request by RobCattonKPMG**
## Description

Enabling the netex converter to create flat fare priced by distance netex.

## Testing instructions

Use the make file to generate all and validate all.

### CODE REVIEW

File: repos/fdbt-netex-output/src/netex-convertor/generateAll.test.ts [hunk 1]
```diff
@@ -52,6 +52,9 @@ describe('generateAll', () => {
 
         const netexGen = await netexGenerator(ticket, operatorData);
         const generatedNetex = await netexGen.generate();
+        // if (fileName === 'flatFareWithDistances.json') {
```
Comment (NitishAroraKPMG) [inline repos/fdbt-netex-output/src/netex-convertor/generateAll.test.ts:55]: remove the commented code

File: repos/fdbt-netex-output/src/netex-convertor/netexGenerator.ts [hunk 2]
```diff
@@ -352,6 +354,12 @@ const netexGenerator = async (ticket: Ticket, operatorData: Operator[]): Promise
     const updatePriceFareFrame = (priceFareFrame: NetexObject): NetexObject => {
         const priceFareFrameToUpdate = { ...priceFareFrame };
```
Comment (NitishAroraKPMG) [inline repos/fdbt-netex-output/src/netex-convertor/netexGenerator.ts:356]: const isPricedByDistance = 'pricingByDistance' in ticket.products[0];

File: repos/fdbt-netex-output/src/netex-convertor/netexGenerator.ts [hunk 3]
```diff
@@ -352,6 +354,12 @@ const netexGenerator = async (ticket: Ticket, operatorData: Operator[]): Promise
...
 
+        if (!('pricingByDistance' in ticket.products[0])) {
```
Comment (NitishAroraKPMG) [inline repos/fdbt-netex-output/src/netex-convertor/netexGenerator.ts:357]: if (!isPricedByDistance) {

File: repos/fdbt-netex-output/src/netex-convertor/netexGenerator.ts [hunk 4]
```diff
@@ -370,12 +378,22 @@ const netexGenerator = async (ticket: Ticket, operatorData: Operator[]): Promise
         tariff.fareStructureElements.FareStructureElement = fareStructuresElements;
 
         if (isFlatFareType(ticket)) {
+            const isPricedByDistance = 'pricingByDistance' in ticket.products[0];
```
Comment (NitishAroraKPMG) [inline repos/fdbt-netex-output/src/netex-convertor/netexGenerator.ts:381]: we can move this up

File: repos/fdbt-netex-output/src/netex-convertor/netexGenerator.ts [hunk 5]
```diff
@@ -352,6 +354,12 @@ const netexGenerator = async (ticket: Ticket, operatorData: Operator[]): Promise
     const updatePriceFareFrame = (priceFareFrame: NetexObject): NetexObject => {
         const priceFareFrameToUpdate = { ...priceFareFrame };
 
+        if (!('pricingByDistance' in ticket.products[0])) {
+            delete priceFareFrameToUpdate.geographicalUnits;
+            delete priceFareFrameToUpdate.tariffs.Tariff.GeographicalUnitRef;
+            delete priceFareFrameToUpdate.tariffs.Tariff.geographicalIntervals;
+        }
```
Comment (NitishAroraKPMG) [inline repos/fdbt-netex-output/src/netex-convertor/netexGenerator.ts:361]: else {

tariff.geographicalIntervals.GeographicalInterval = getGeographicalIntervals(
                    ticket.products[0] as PriceByDistanceProduct,
                );

}

File: repos/fdbt-netex-output/src/netex-convertor/netexGenerator.ts [hunk 6]
```diff
@@ -370,12 +378,22 @@ const netexGenerator = async (ticket: Ticket, operatorData: Operator[]): Promise
...
-                ref: 'fxc:flat',
+                ref: typeOfTariffRef,
             };
 
-            tariff.TariffBasis = { $t: 'flat' };
+            tariff.TariffBasis = { $t: tariffBasis };
+
+            if (isPricedByDistance) {
+                tariff.geographicalIntervals.GeographicalInterval = getGeographicalIntervals(
+                    ticket.products[0] as PriceByDistanceProduct,
```
Comment (NitishAroraKPMG) [inline repos/fdbt-netex-output/src/netex-convertor/netexGenerator.ts:394]: can we move this up for clarity about what we are doing with Geographical unit in pricedbydistance

File: repos/fdbt-netex-output/src/netex-convertor/period-tickets/periodTicketNetexHelpers.ts [hunk 7]
```diff
@@ -423,6 +432,103 @@ const getFlatFareList = (
         });
     });
 
+export const getGeographicalIntervalPrices = (distanceBands: DistanceBand[]): GeographicalIntervalPrice[] => {
+    const prices = distanceBands.flatMap(band => {
+        const { distanceTo, distanceFrom } = band;
+        let currentDistanceBand = 0;
+        const numberOfPricesToGenerate = Number(distanceTo) - Number(distanceFrom);
```
Comment (NitishAroraKPMG) [inline repos/fdbt-netex-output/src/netex-convertor/period-tickets/periodTicketNetexHelpers.ts:439]: the distanceTo can be *Max* for the last row

File: repos/fdbt-netex-output/src/netex-convertor/generateAll.test.ts (hunk 1 above)
Comment (RobCattonKPMG) [reply repos/fdbt-netex-output/src/netex-convertor/generateAll.test.ts:55]: good catch

File: repos/fdbt-netex-output/src/netex-convertor/netexGenerator.ts (hunk 4 above)
Comment (RobCattonKPMG) [reply repos/fdbt-netex-output/src/netex-convertor/netexGenerator.ts:381]: But we only want to check it if its a flat fare ticket.

File: repos/fdbt-netex-output/src/netex-convertor/netexGenerator.ts (hunk 2 above)
Comment (RobCattonKPMG) [reply repos/fdbt-netex-output/src/netex-convertor/netexGenerator.ts:356]: Ah you mean move the boolean to a variable

File: repos/fdbt-netex-output/src/netex-convertor/netexGenerator.ts (hunk 5 above)
Comment (RobCattonKPMG) [reply repos/fdbt-netex-output/src/netex-convertor/netexGenerator.ts:361]: im not sure i understand what you mean here?

File: repos/fdbt-netex-output/src/netex-convertor/period-tickets/periodTicketNetexHelpers.ts (hunk 7 above)
Comment (RobCattonKPMG) [reply repos/fdbt-netex-output/src/netex-convertor/period-tickets/periodTicketNetexHelpers.ts:439]: it already is.

File: repos/fdbt-netex-output/src/netex-convertor/generateAll.test.ts (hunk 1 above)
Comment (alainajaffer) [reply repos/fdbt-netex-output/src/netex-convertor/generateAll.test.ts:55]: this is still here

**Review by alainajaffer — APPROVED** [review]

**Pull request merged**